src/java/org/apache/cassandra/db/tries/RangesCursor.java

```diff
@@ -0,0 +1,324 @@
+/// forward as well as backward order. This in turn makes it possible to have simple range trie and intersection
+/// implementations where the representations of range trie sets do not differ when iterated in the two directions.
+///
+/// Thes types of ranges are actually preferable to us, because we use prefix-free keys with terminators that leave
+/// room for greater- and less-than positions, and at prefix nodes we store metadata applicable to the whole branch.
+///
+/// The ranges are specified by passing a sequence of boundaries, where each even boundary is the opening boundary, and
+/// the odd ones are the closing boundary for a range. The boundaries must be in order and cannot overlap, but it is
+/// possible to repeat a boundary, e.g.
+/// - `[a, a] == point a`: an even number of repeats starting at an even position specifies a point
+///    (i.e. the set of all prefixes and all descendants of that key)
+/// - `[a, b, b, c] == [a, c]`: an even number of repeats which starts at an odd position is ignored
+/// - `[a, a, a, b] == [a, b]`: an odd number of repeats is the same as a single copy
+///
+/// If the first boundary is null, the range is open on the left, and if the last non-null boundary is on an even
+/// position (i.e. if the array has an even length and its last boundary is null, or if that last boundary is cut off),
+/// the range is open on the right:
+/// - `[null, a]` covers all keys smaller than `a`, plus the prefixes and descendants of `a`
+/// - `[a, null]` or `[a]` covers all keys greater than `a`, plus the prefixes and descendants of `a`
+class RangesCursor implements TrieSetCursor
+{
+    private final ByteComparable.Version byteComparableVersion;
+    private final Direction direction;
+
+    /// The current index in the boundaries array. This is the input that will be advanced on the next [#advance] call.
+    /// When the current boundary is exhausted, it will advance (in the direction of the cursor). When it surpasses
+    /// [#completedIdx], the cursor is exhausted.
+    private int currentIdx;
+    /// The index at which the ranges end. Depending on the direction of iteration, the end or start of the array,
+    /// adjusted to remove null boundaries.
+    private final int completedIdx;
+    /// The next byte for all boundaries.
+    int[] nexts;
+    /// The depth (processed number of bytes) for all boundaries.
+    int[] depths;
+    /// Byte sources producing the rest of the bytes of the boundaries.
+    ByteSource[] sources;
+    /// The currently reached depth (reported to the user). This is usually `depths[currentIdx] - 1`.
+    int currentDepth;
+    /// The current incoming transition.
+    int currentTransition;
+    /// Current range state, returned by [#state].
+    RangeState currentState;
+
+    public RangesCursor(Direction direction, ByteComparable.Version byteComparableVersion, ByteComparable... boundaries)
```

Comment:
This does a lot for a constructor.
Maybe this could be a static `RangesCursor.from` method that calls the private constructor ?